### PR TITLE
dev-lang/ruby: 2.4.5-r1 for OpenSSL 1.1 "disable-deprecated"

### PR DESCRIPTION
**Pull request by candrews**
Tested by running:
`USE="rubytests" FEATURES="test" ebuild ruby-2.4.5-r1.ebuild test`
Everything passes.

Closes: https://bugs.gentoo.org/614760
Package-Manager: Portage-2.3.52, Repoman-2.3.12
Signed-off-by: Craig Andrews <candrews@gentoo.org>

@Polynomial-C should interest you

### CODE REVIEW

File: dev-lang/ruby/files/2.4/012-openssl_1.1.patch [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+-    ERR_remove_state(0);
+-    ERR_free_strings();
+ #endif
+ 
+     /*
+@@ -1149,7 +1138,11 @@ Init_openssl(void)
+     /*
+      * Version of OpenSSL the ruby OpenSSL extension is running with
+      */
++#if !defined(LIBRESSL_VERSION_NUMBER) && OPENSSL_VERSION_NUMBER >= 0x10100000
```
Comment (steils) [inline dev-lang/ruby/files/2.4/012-openssl_1.1.patch]: `+#if OPENSSL_VERSION_NUMBER >= 0x10100000L || LIBRESSL_VERSION_NUMBER >= 0x2070000fL`

File: dev-lang/ruby/files/2.4/012-openssl_1.1.patch [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+diff --git a/ext/openssl/ossl_engine.c b/ext/openssl/ossl_engine.c
+index d69b5dca..5ca0d4ca 100644
+--- a/ext/openssl/ossl_engine.c
++++ b/ext/openssl/ossl_engine.c
+@@ -46,13 +46,25 @@ VALUE eEngineError;
+ /*
+  * Private
+  */
+-#define OSSL_ENGINE_LOAD_IF_MATCH(x) \
++#if !defined(LIBRESSL_VERSION_NUMBER) && OPENSSL_VERSION_NUMBER >= 0x10100000
```
Comment (steils) [inline dev-lang/ruby/files/2.4/012-openssl_1.1.patch]: `+#if OPENSSL_VERSION_NUMBER >= 0x10100000L || LIBRESSL_VERSION_NUMBER >= 0x2070000fL`

File: dev-lang/ruby/files/2.4/012-openssl_1.1.patch [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+-    OSSL_ENGINE_LOAD_IF_MATCH(openssl);
++    OSSL_ENGINE_LOAD_IF_MATCH(openssl, OPENSSL);
+     rb_warning("no such builtin loader for `%"PRIsVALUE"'", name);
+     return Qnil;
+ #endif /* HAVE_ENGINE_LOAD_BUILTIN_ENGINES */
+@@ -160,7 +172,9 @@ ossl_engine_s_load(int argc, VALUE *argv, VALUE klass)
+ static VALUE
+ ossl_engine_s_cleanup(VALUE self)
+ {
++#if defined(LIBRESSL_VERSION_NUMBER) || OPENSSL_VERSION_NUMBER < 0x10100000
```
Comment (steils) [inline dev-lang/ruby/files/2.4/012-openssl_1.1.patch]: `+#if OPENSSL_VERSION_NUMBER < 0x10100000L || (defined(LIBRESSL_VERSION_NUMBER) && LIBRESSL_VERSION_NUMBER < 0x2070000fL)`

File: dev-lang/ruby/files/2.4/012-openssl_1.1.patch [hunk 4]
```diff
@@ -0,0 +1,339 @@
...
+diff --git a/ext/openssl/ossl_ssl.c b/ext/openssl/ossl_ssl.c
+index 8e3c0c42..d32a299c 100644
+--- a/ext/openssl/ossl_ssl.c
++++ b/ext/openssl/ossl_ssl.c
+@@ -379,7 +379,7 @@ ossl_call_session_get_cb(VALUE ary)
+ 
+ /* this method is currently only called for servers (in OpenSSL <= 0.9.8e) */
+ static SSL_SESSION *
+-#if OPENSSL_VERSION_NUMBER >= 0x10100000L && !defined(LIBRESSL_VERSION_NUMBER)
++#if OPENSSL_VERSION_NUMBER >= 0x10100000 && !defined(LIBRESSL_VERSION_NUMBER)
```
Comment (steils) [inline dev-lang/ruby/files/2.4/012-openssl_1.1.patch]: Why change `0x10100000L` => `0x10100000`?
Anyway,
`+#if OPENSSL_VERSION_NUMBER >= 0x10100000L || LIBRESSL_VERSION_NUMBER >= 0x2070000fL`

**Comment by steils** [comment]
Oh, I see, it's written exactly like in the upstream, so the upstream Ruby doesn't fully support LibreSSL>=2.7.0. Then maybe it's not worth it to accept my changes, but that's up to you

**Comment by candrews** [comment]
> LibreSSL>=2.7.0 provides OPENSSL_init_ssl(). If you don't want to force LibreSSL users to use deprecated functions it could be something like

So does this work with LibreSSL>=2.7.0, it just requires deprecated functions to be enabled (and it seems that libressl in Gentoo does have deprecated functions enabled)?

>Oh, I see, it's written exactly like in the upstream, so the upstream Ruby doesn't fully support LibreSSL>=2.7.0. Then maybe it's not worth it to accept my changes, but that's up to you

If this PR does work with LibreSSL>=2.7.0 as it is in Gentoo today, then IMHO, it's best to not accept your changes and stick to upstream. (I'm not the maintainer of dev-lang/ruby, so this is just my opinion - I'll happily update this PR to be whatever the maintainer wants if it means getting this PR accepted and the bug fixed).

On another note, you should probably submit a PR to Ruby so that your LibreSSL changes get into upstream eventually :)

**Comment by steils** [comment]
I didn't test it with LibreSSL>=2.7.0 (I live with 2.6.5), but I'm pretty sure it should work (see https://github.com/ruby/openssl/pull/193/commits/75de15ddcdab6efe7faf3ca1f6b5c6e5b6ba57cc BTW). The proposed chages were about that LibreSSL-2.7.0 provides many functions of the OpenSSL 1.1 API (AFAIK it's almost fully compatible), so it's not nice to use old API.

> IMHO, it's best to not accept your changes and stick to upstream
> ...
> On another note, you should probably submit a PR to Ruby so that your LibreSSL changes get into upstream eventually :)

I agree with both statements. If I had noticed that these changes just repeat the upstream patch, I wouldn't probably review this PR.

**Comment by graaff** [comment]
@candrews thanks for updating the patch. It's now merged.

@steils this is just a backport patch where upstream is unlikely to backport this to 2.4 themselves given the ruby 2.4 lifetime. We try to stick only to upstream patches, so best to address libressl >= 2.7.0 upstream first.

**Pull request closed without merging**